### ISSUE TITLE
Ideas for improving playtime tracking

### ISSUE THREAD

**Issue by unfa** [lutris/lutris#3857]
When using Steam I've noticed it's playtime tracking is flawed in numerous ways, which meant my games sometimes have muhc more playtime on record that I have actually played them. I wonder if Lutris could improve upon that.

1. Playtime keeps rolling if you suspend or hibernate the system while the game is running.
2. Playtime keeps rolling even when you've minimized the game or switched to another window.

If it's possible to track the windows spawned by game's processes to only count playtime when the game's window is in focus - that'd solve switching away from a paused game.

Tracking if there's player input could help as well - if someone pauses the game and goes for lunch or mowing their lawn they will produce no input for a long period of time. Lutris could track if there's been any inputs and write down the last time an input was given. If the time since the latest input crosses a threshold (say 15 minutes, to account for long cutscenes etc.) the time rolls back by the 15 minutes and is paused until new user input is detected.

And also - Lutris could track if the system is actually operating in case of a hard crash, power loss or suspension/hibernation.
I think an error prone and platform independent way would be to write the current hardware clock time to a file on disk every minute (hoping to avoid user altering the system clock). Later the file could be inspected to see if there's any holes in there - it will obviously not be able to write to the file if the system crashes, and using the realtime clock should provide a way to spot these moments and prevent game's playtime to roll when the PC is suspended, something has crashed or power was cut in the entire country.

What do you think? Is there something I missed? Are these feasible features to implement?

**Comment by strycore** [comment]
Those are some pretty advanced features and I wonder if any of them will be possible with Wayland in mind. Right now, Lutris only monitors process names and PIDs. 

Monitoring window position, focus and inputs is something I would have liked to do but it doesn't seem compatible with how things are currently headed.

This may be suitable to implement in gamescope, if there's a way to have gamescope report back the total running time for a game. Another possibility would be to implement it in mangohud.

That said, the idea of having the playtime constructed with increments of time would be a better way than our current way of subtracting the start time from the end time. This would solve cases where the computer is hibernated / suspended.
This would also allow keeping the playtime up to date in the lutris UI while the game is running.

**Comment by ntninja** [comment]
FYI, the case of hibernation/suspension is trivial to solve by switching to a different clock source: Python’s [`time.monotonic()`](https://docs.python.org/3.9/library/time.html?highlight=time#time.monotonic) function uses `clock_gettime(CLOCK_MONOTONIC)` to return seconds since boot excluding any time the system was suspended – by extension this also means any changes in the system clock settings are ignored as well. (Note that POSIX only specifies that the clock must be monotonically increasing, hence the name, from an unspecified point in the past and must be represented in seconds and nanoseconds, the additional semantics mentioned above are guaranteed by the Linux kernel only.)

Points 2 and 3 are indeed tricky – ideally games would just tell Lutris these values, but I know how likely *that* is…

**Comment by teotikalki** [comment]
The hibernate thing is really noticeable to me; some update to League of Legends earlier this year made it so that when you close the actual game client the tray/launcher bit sticks around, and I often play League right before I go to bed. I've often turned my computer on the next day and watched my playtime jump by 20+ hours because of this.

**Comment by Ileca** [comment]
I am reliant on this feature but the time being recorded while the window lost the focus makes the times highly unreliable. Worse, it makes you unable to enjoy your game properly as you stop yourself from taking breaks or multitask (I play a lot of games without fullscreen) because you know the time will be more and more inaccurate.

I thought Wayland didn't allow you to know about other processes for obvious security reasons but shouldn't knowing if your own window has the focus or not be trivial? I don't know, I am just asking.

**Comment by strycore** [comment]
>  shouldn't knowing if your own window has the focus or not be trivial?

yes, but the Lutris window and the game window are 2 separate things. We could possibly pause the playtime if get the focus back to the Lutris window but we have no way of knowing when the focus is given back to the game.

**Comment by Ileca** [comment]
Couldn't you store the time in the game which should know if it has the focus and send it back to Lutris before it is shut down? You still take the time from Lutris in case something bad happen and the time is lost.

**Comment by strycore** [comment]
That would require to be added to games themselves and follow some open standards of reporting playtime.  The only process capable of knowing the game's window status reliably is the game executable which is a proprietary binary most  of the time. Maybe Gamescope can implement such a thing.
